### ISSUE TITLE
Another question about the uw.conditions.DirichletCondition?

### ISSUE THREAD

**Issue by tyszwh** [underworldcode/underworld2#193]
Another question about the uw.conditions.DirichletCondition, I can't understand the parameter indexSetsPerDof = (iwalls, iwalls) or other settings. Is there a more useful detail guide for  possible boundary conditions?

I see the issues have mentioned there have a guide for boundary condition. But I did not find it.

**Comment by jmansour** [comment]
As per #173, some of this information has been removed and will be rewritten soon.  You can access the old version here:

https://github.com/underworldcode/underworld2/blob/61530e3ad18f2b40c4107b3954285fcb17bf8e11/docs/user_guide/06_Systems.ipynb

Also, the docstring for this class might be useful to you:
https://github.com/underworldcode/underworld2/blob/master/underworld/conditions/_dirichletCondition.py#L34

Basically, for scalar fields (such as a temperature), there is only a single degree of freedom (Dof) at each node, so you will only provide a single index set which specifies which nodes you wish to flag as Dirichlet.  For vector fields (such as velocity), there will be 2 or 3 degrees of freedom per node, and you must provide an index set for *each* Dof (v_x, v_y, v_z) which specifies which node/dof you wish to flag as Dirichlet. 

Hope this helps!

**Comment by tyszwh** [comment]
Thanks for your explanation. It really helps. 
But I still have question on the velocity boundary. Such as we give a velocity on left walls and right walls (iWalls). 
```
for index in mesh.specialSets["MinI_VertexSet"]:
    velocityField.data[index] = [-1, 0.]
for index in mesh.specialSets["MaxI_VertexSet"]:
    velocityField.data[index] = [ 1, 0.]
```
indexSetsPerDof = (iWalls, base) (example 209) means the iWalls  fix with 1 or 01 in x direction and the basewalls fix with 0 in y direction. Is it right? So what the topwalls do now? (If I just give the indexSetsPerDof = (jWalls, jWalls), what iwalls do in this situation.)
 The result show that leftwalls and rightwalls moved with velocity directions. If I dont want the leftwalls and rightwalls moves with velocity (such as the material flows out?) . what should I do?

 It different with underworld1. In underworld1, we just set the left walls and right walls in x direction and give a opposite velocity value. Whether he wall move or fix is depend on the movemesh setting. So I felt confused with it.
 
Additionally, If we want to give velocity boundary as function of X ,Y or Z direction (the "func" in underworld1 ).  Such as the half left walls is -0.5 ,half left walls is 0.5 (half left walls means the walls was  split horizontal or vertical with half area )
What should I do?

**Comment by jmansour** [comment]
In example 2_09, yes you are correct about the iWalls and the basewall.  The topwall has no Dirichlet (or other) condition set upon it, and it therefore it will implicitly have a Neumann type condition which imposes a zero normal stress requirement.

If you do not wish for the mesh to move, simply remove the `with deform_mesh` blocks. Note that your particle will leave the domain, and therefore you will need to set the [particleEscape parameter](http://underworld2.readthedocs.io/en/latest/underworld.swarm.html#underworld.swarm.Swarm) to true when you create your swarm (if particles are leaving), and/or add a population control mechanism (as per example 2_09). 

To set more complex boundary conditions, simply loop over the boundary nodes within python.  Something like this perhaps:
```python
for index in mesh.specialSets["MinI_VertexSet"]:
    coord = mesh.data[index]
    if coord[1] > domainHeight/2.:
        velocityField.data[index] = [0.5,0.]
    else:
        velocityField.data[index] = [-0.5,0.]

```

**Comment by tyszwh** [comment]
Thanks for your help.  It is a perfect explantion.

**Comment by jmansour** [comment]
No probs!  I'll close this ticket, but feel free to reopen it if you have further questions.

**Comment by tyszwh** [comment]
Ok , I have met the other questions.
1 I want get a displacement field of a surface which include X, Y,Z directions. I have added radom swarm particle for a entire surface.  And I rum the model, get the result. The result can include the position of X, Y,Z for swarms. If I use the position of timestep2 minus the timestep1. I can get the displacement. right?  or I use  the velocity filed of swarm particle to multipy the dt(timestep2-timestep1) to get the displacement field. But I found the number of swarm partilcle changed in every timestep. The swarm flow out the mesh domain.So I Cant do it like this way.  Getting the displacement filed is to obtain the tectonic movent Which can be used in surface process model. I want found the method for both underworld1 and underworld2.What should I do?
 I also use the velocity field data in output with fixed mesh node(not the passive swarm point tracer).But I calculate the velocity field data in every tiemstep with dt. I got a high value.I thought this way is a wrong method because of the value of velocity field at fixed mesh node. 

2 If I want the boundary condition varied with time in uncerworld2. What should I do with define the boundary condition value.

**Comment by jmansour** [comment]
1.  The way to do this would be to record the position on the particles as a swarm variable.  So create a swarm for the displacement particles (make sure this is a different swarm to your material swarm!).  Add the particles where you want them.  Add a new swarm variable which you will use to record the original position.  Advect the displacement swarm, and then diff against the particles original positions (as recorded on each particle).   Let me know if this isn't clear. 

2. Because in UW2 the time-stepping loop is completely exposed, you can simply change the values on the boundary node/dofs whenever you require.  So remember that using the DirichletCondition object, you are simply flagging nodes/dofs as BCs... but not actually setting the BC value on the node/dof.  Setting the value can be done at *any* time of the simulation... So say you run for 10 timesteps, you can then switch the BCs as necessary, and then continue the timestep loop as necessary.  Just be careful to obtain a new max_dt *after* you have changed your BC values. 

Also, as this ticket is closed, perhaps just start a new ticket for further questions, unless they relate directly to the original question of this ticket (in which case reopen this ticket).

**Comment by lmoresi** [comment]
Suggest we think of a way to answer "modelling" questions and share the answers.

These are not issues with the code or with the documentation but the community will benefit from hearing the answers.

Sent from my iPhone

On 18 Feb 2017, at 08:34, jmansour <notifications@github.com<mailto:notifications@github.com>> wrote:


  1.  The way to do this would be to record the position on the particles as a swarm variable. So create a swarm for the displacement particles (make sure this is a different swarm to your material swarm!). Add the particles where you want them. Add a new swarm variable which you will use to record the original position. Advect the displacement swarm, and then diff against the particles original positions (as recorded on each particle). Let me know if this isn't clear.

  2.  Because in UW2 the time-stepping loop is completely exposed, you can simply change the values on the boundary node/dofs whenever you require. So remember that using the DirichletCondition object, you are simply flagging nodes/dofs as BCs... but not actually setting the BC value on the node/dof. Setting the value can be done at any time of the simulation... So say you run for 10 timesteps, you can then switch the BCs as necessary, and then continue the timestep loop as necessary. Just be careful to obtain a new max_dt after you have changed your BC values.

Also, as this ticket is closed, perhaps just start a new ticket for further questions, unless they relate directly to the original question of this ticket (in which case reopen this ticket).

**Comment by jmansour** [comment]
Problem solved? 😀

**Comment by tyszwh** [comment]
Oh,no, I still need to try in underworld1. Lately, I will try the undrworld2. The underworld2 run slowly with a low resolution model (8*8*8),but the underworld1 fast.  Something wrong with my setting?

**Comment by tyszwh** [comment]
The main problem in underworld1 for this problem is: I read the passive tracer from swarm output in paraview or maybe from underworld swarm output. But ，the number of swarm tracer which I defined for recording the position is varied. Such as the 1000 in timestep 0(I defined), maybe 800 in timestep 10(at least I observed the data varied from paraview show like this ). So I cant calculate the displacement from time step 10 minus the timestep 0 because we have different numbers of swarm particle.
And how to create a user defined swarmvariable in underworld1(you mentioned in the method)? I know It is easy to do in underwolrd2.
